### DISCUSSION TITLE
Ideas, Questions, Mockups, etc. for Minesweeper 2.0

### DISCUSSION THREAD

**Discussion by jonasgeiler** [Skayo/TI-Nspire-Minesweeper#6] [Ideas]
This discussion should be used for any ideas, questions, mockups and similar for the overhaul of the game.
You can find the task managment board [here](https://github.com/Skayo/TI-Nspire-Minesweeper/projects/3).

[the remark this answers is not in the document]
**Reply by jonasgeiler** [reply]
That certainly would be very cool! If it has the same "look" as the rest of the UI/Game it would definitely look a lot better than the simple menus.

You could work on this feature on a new branch an make a PR at the end so I can check it out!

Thank you for putting so much work into this!

[the remark this answers is not in the document]
**Reply by jonasgeiler** [reply]
How are your UI/UX skills? Should I try to make a little mockup for the interface?

[the remark this answers is not in the document]
**Reply by jonasgeiler** [reply]
Looks pretty good!
The only thing I'd change is having more options in the Pause Menu.
So in the running game you press ESC -> Pause Menu with
- Resume
- Settings
- Help
- Restart (with confirmation)
- Exit to Main Menu (or just "Exit") (with confirmation)

The Main Menu could also have a "Help" option.

**Comment by jonasgeiler** [comment]
I found this yesterday: https://tcrf.net/Proto:Minesweeper_(Windows,_1990)/Mine_2.9
It shows some images from the Original Game Prototype, with a settings menu.
I will probably base the mockups on the design they had back then with the buttons and so on.
Not sure if I have the time for it today tho.

**Comment by jonasgeiler** [comment]
Another thing which could be nice is some kind of key remapping. So you can set your own buttons for moving the cursor, setting flags, etc.

Also I thought about some simple localization using the Locale API from TI Nspire. This would only require a table with  the translated strings for each language. I could provide a German translation for the game and it would just fallback to English if the user's language doesn't exist.

And, at last, what would you think about cursor control? So you can move around the cursor and click on the tiles instead of navigating with the keys? Do you think that's a lot of work? 

And don't worry I definitely want to help you with some features! I just hope my Lua knowledge is still up-to-date haha

I'll start with the mockups later today and maybe set up a simple GitHub Project to manage our plans and stuff.

I hope I'm not turning this little game into something way too complex 😅
But on the other side this is one of my biggest projects according to the amount of downloads lol

**Reply by jonasgeiler** [reply]
Not a problem! Just do it at your own pace

**Comment by jonasgeiler** [comment]
Maybe you've already done it but if you are interested you can download the original minesweeper from [here](https://minesweepergame.com/download/windows-xp-minesweeper.php) and reverse engineer/compare the behavior and stuff.
I recently noticed that the smiley doesn't have the "scared" state while clicking on a tile

**Edit:**
Just found the original source code haha
https://github.com/cryptoAlgorithm/nt5src/tree/daad8a087a4e75422ec96b7911f1df4669989611/Source/XPSP1/NT/shell/osshell/ep/winmine

**Comment by jonasgeiler** [comment]
As you can see there is now a Kanban Board [here](https://github.com/Skayo/TI-Nspire-Minesweeper/projects/3).
We'll use that to keep track of all the tasks. It should be rather self-explaining, even if you haven't used a Kanban Board before. Otherwise just ask 😊
Feel free to add your own cards too!

**Comment by jonasgeiler** [comment]
What do you think about drawing the borders manually? If you check the source code of the original game you can see that they don't use separate images for the borders, but draw them with this function [here](https://github.com/cryptoAlgorithm/nt5src/blob/daad8a087a4e75422ec96b7911f1df4669989611/Source/XPSP1/NT/shell/osshell/ep/winmine/grafix.c#L420).
This way we could reduce file size and also simplify reusing the border drawing for menus and such.

**Reply by alexcoder04** [reply]
That would make sense

**Comment by jonasgeiler** [comment]
I think we should take the [Original Help Files](https://github.com/cryptoAlgorithm/nt5src/tree/daad8a087a4e75422ec96b7911f1df4669989611/Source/XPSP1/NT/shell/osshell/ep/winmine/help) (.rtf format) and try to convert them into a .tns document / Luna-compatible XML.
Then we provide a separate "Help.tns" next to the "Minesweeper.tns", and if the user needs it, they upload it to their calculator. Otherwise they only need to upload the game itself.
It could look a bit like the "Getting Started.tns" which exists on the calculator by default (we could even insert images).

What do you think?

**Reply by alexcoder04** [reply]
I think it would make more sense to insert the help document as an extra page into the .tns file rather than putting it into an own file.

**Reply by jonasgeiler** [reply]
The Problem is it would be A LOT of text/images/etc. so either we don't use the original help files or we make it separate

**Comment by jonasgeiler** [comment]
Did you put in a lot of work for the UI already?
Because I think I would actually keep the tool palette and use actual UI elements for specific stuff only, like in the original game.

So we would have a tool palette which looks a bit like this:
![Tool Palette](https://user-images.githubusercontent.com/10259118/147792502-cd87e1d9-5278-4fe6-bb12-9a84f90d2cc9.png)

And when you select "Custom...", it opens a window like this:
![Custom Window](https://user-images.githubusercontent.com/10259118/147792526-bf46c257-b9c1-4edc-ae5e-19d24953bed9.png)

And when you select "Best Times..." (calling it "Highscores" is actually wrong, I think), it looks like this:
![Highscores](https://user-images.githubusercontent.com/10259118/147792574-e6c2131e-dbb2-4b3e-b8cf-22976304c3aa.png)

And when you set a new best time, it shows something like this:
![New Best Time](https://user-images.githubusercontent.com/10259118/147792712-a4ca1d25-f5a2-4e18-8609-87c0bd25791f.png)

The pause screen could have the same look as the main game, but instead of the tiles/minefield, it just shows a text with "Paused".
Let me know what you think about this. I prefer it like this, because it immediately starts into the game and doesn't show an annoying Main Menu first. Also you don't have to make a whole pause menu because the tool palette acts as both a main menu AND pause menu, you know?

**Reply by alexcoder04** [reply]
I think this is overall a good idea. I looked up if there are Lua bindings for the native TI dialog boxes and UI elements, like the ones that are used for picking the device's language (you can use some of them in BASIC), but from my research they are not available in Lua scripts.

**Reply by alexcoder04** [reply]
I am asking myself, what is the most efficient/elegant way to implement such popup windows: drawing a rectangle on top of the game with the text and the buttons, save the buttons position and check if mouse position matches these coordinates on click? Or writing a tiny "GUI toolkit" which handles that stuff?

**Reply by alexcoder04** [reply]
> Also you don't have to make a whole pause menu because the tool palette acts as both a main menu AND pause menu, you know?

The timer goes on if you open the tool palette and can't think of any way to run something (stop timer, draw "paused" text) if the tool palette is opened

**Reply by jonasgeiler** [reply]
> I am asking myself, what is the most efficient/elegant way to implement such popup windows: drawing a rectangle on top of the game with the text and the buttons, save the buttons position and check if mouse position matches these coordinates on click? Or writing a tiny "GUI toolkit" which handles that stuff?

I think a tiny GUI toolkit would be really cool! You could base it upon the GUI Engine from [nSpaint](https://www.ticalc.org/archives/files/fileinfo/447/44726.html) and customize it a bit so the dialogs look like this:
![Dialog](https://user-images.githubusercontent.com/10259118/148284274-f823bd0b-85f0-424a-ada6-c53fbc3e945a.png)
(no need for the icon at the top-left, and the buttons on the top-right tho)


> The timer goes on if you open the tool palette and can't think of any way to run something (stop timer, draw "paused" text) if the tool palette is opened

I think we should use the ESC-key or P-key for pausing the game anytime. And if the user wants to change level or something they can open the tool palette and don't need an extra "Pause Menu". You know?
And if you open the tool palette in the middle of the game the timer just continues.

**Reply by alexcoder04** [reply]
> I think a tiny GUI toolkit would be really cool! You could base it upon the GUI Engine from nSpaint and customize it a bit so the dialogs look like this:

I had a look at the nSpaint GUI engine and did the "Best times" table using it, would just need to adjust the design and add the "enter name" screen.
However the fact came to my mind, that nSpaint does not have an open source license, so it's technically copyrighted. So we should ask the dev if we use his code, shouldn't we?

**Reply by jonasgeiler** [reply]
> However the fact came to my mind, that nSpaint does not have an open source license, so it's technically copyrighted. So we should ask the dev if we use his code, shouldn't we?

You're right... Didn't really think of that. The authors email is in the Readme.txt of nSpaint, so I'll be writing an email later. Let's hope he responds

**Comment by jonasgeiler** [comment]
Oh btw I think we should change the background color to `RGB(59, 110, 165)` so it matches the default Windows 2000 wallpaper!

**Comment by alexcoder04** [comment]
There are two limitations that I noticed today:

1. Custom field size does not make a lot of sense on the small calculator screen, it would be only one or two columns different from the default levels.
2. There is not much that can be done to the tool palette, because there is no option to make an item without submenus or an item with a checkmark, like in the original game.

**Reply by jonasgeiler** [reply]
> 1. Custom field size does not make a lot of sense on the small calculator screen, it would be only one or two columns different from the default levels.

Yeah true I guess the limit would be 2x2 to 10x18. Let's make this like really low priority and only implement it if it's like less than an hour of work or something.

> 2\. There is not much that can be done to the tool palette, because there is no option to make an item without submenus or an item with a checkmark, like in the original game.

Is it possible to use like `✓`, `✔` or `√` in front of the text in the tool palette as like a work-around?

**Reply by alexcoder04** [reply]
> Is it possible to use like ✓, ✔ or √ in front of the text in the tool palette as like a work-around?

I already thought about something like that, but that would require a submenu for ticking it on/off, which looks kinda weird. However, it would probably come closest to the original.

**Reply by jonasgeiler** [reply]
> I already thought about something like that, but that would require a submenu for ticking it on/off, which looks kinda weird. However, it would probably come closest to the original.

Just noticed I missed the "no option to make an item without submenus" part... Damn. Guess we'll have to keep it as it is right now...
